Cannot start integration test container using Podman (rootless) and bind mounts

### Describe the bug

Hello,
when enabling integration testing together with `quarkus.container-image.build=true`, the `quarkus-container-image-podman` dependency and a mount like `quarkus.test.container.volume-mounts."/path/to/quarkus-podman-issue/io/github/gtache/key-store.p12"=/deployments/key-store.p12`, integration testing fails when starting the container with the message (in `target/quarkus.log`):
```{"msg":"exec container process `/opt/jboss/container/java/run/run-java.sh`: Permission denied","level":"error","time":"2025-07-25T08:02:18.425431Z"}```

(On my actual project I get ```{"msg":"exec container process `/usr/local/s2i/run`: Permission denied","level":"error","time":"2025-07-24T16:05:53.247336Z"}```, but the problem is the same)

The command used by Quarkus to run the container is 
```
podman run --name quarkus-integration-test-uWlxL -i --rm --user 2803:110 --userns=keep-id -p 47663:47663 -p 46835:46835 -v /path/to/quarkus-podman-issue/src/test/resources/io/github/gtache/key-store.p12:/deployments/key-store.p12:z --net=quarkus-integration-test-DEvAX --env QUARKUS_LOG_CATEGORY__IO_QUARKUS__LEVEL=INFO --env QUARKUS_HTTP_PORT=47663 --env QUARKUS_HTTP_SSL_PORT=46835 --env TEST_URL=http://localhost:0 --env QUARKUS_PROFILE=prod --env QUARKUS_TLS_KEY_STORE_P12_PASSWORD=keystore tga/quarkus-podman-issue:1.0.0-SNAPSHOT
```

If we remove the mount and re-run `mvn clean verify`, the command is 
```
podman run --name quarkus-integration-test-uWlxL -i --rm -p 47663:47663 -p 46835:46835 --net=quarkus-integration-test-DEvAX --env QUARKUS_LOG_CATEGORY__IO_QUARKUS__LEVEL=INFO --env QUARKUS_HTTP_PORT=47663 --env QUARKUS_HTTP_SSL_PORT=46835 --env TEST_URL=http://localhost:0 --env QUARKUS_PROFILE=prod --env QUARKUS_TLS_KEY_STORE_P12_PASSWORD=keystore tga/quarkus-podman-issue:1.0.0-SNAPSHOT
```
but obviously this fails due to `Caused by: java.nio.file.NoSuchFileException: /deployments/key-store.p12`

Running the same command as the first manually but without the user and userns args (and the net too because it doesn't exist in this case) works:
```
podman run --name quarkus-integration-test-uWlxL -i --rm -p 47663:47663 -p 46835:46835 -v /path/to/quarkus-podman-issue/src/test/resources/io/github/gtache/key-store.p12:/deployments/key-store.p12:z --env QUARKUS_LOG_CATEGORY__IO_QUARKUS__LEVEL=INFO --env QUARKUS_HTTP_PORT=47663 --env QUARKUS_HTTP_SSL_PORT=46835 --env TEST_URL=http://localhost:0 --env QUARKUS_PROFILE=prod --env QUARKUS_TLS_KEY_STORE_P12_PASSWORD=keystore tga/quarkus-podman-issue:1.0.0-SNAPSHOT
```

```
INFO exec -a "java" java -XX:MaxRAMPercentage=80.0 -XX:+UseParallelGC -XX:MinHeapFreeRatio=10 -XX:MaxHeapFreeRatio=20 -XX:GCTimeRatio=4 -XX:AdaptiveSizePolicyWeight=90 -XX:+ExitOnOutOfMemoryError -Dquarkus.http.host=0.0.0.0 -Djava.util.logging.manager=org.jboss.logmanager.LogManager -cp "." -jar /deployments/quarkus-run.jar 
INFO running in /deployments
__  ____  __  _____   ___  __ ____  ______ 
 --/ __ \/ / / / _ | / _ \/ //_/ / / / __/ 
 -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \   
--\___\_\____/_/ |_/_/|_/_/|_|\____/___/   
2025-07-25 08:15:10,876 INFO  [io.quarkus] (main) quarkus-podman-issue 1.0.0-SNAPSHOT on JVM (powered by Quarkus 3.25.0) started in 1.089s. Listening on: https://0.0.0.0:46835
2025-07-25 08:15:10,878 INFO  [io.quarkus] (main) Profile prod activated. 
2025-07-25 08:15:10,879 INFO  [io.quarkus] (main) Installed features: [cdi, rest, smallrye-context-propagation, vertx]
2025-07-25 08:15:14,282 INFO  [io.quarkus] (main) quarkus-podman-issue stopped in 0.076s
```

So the issue is with these arguments. The user in the container can't access the script to run Quarkus.
I don't think I'm doing anything wrong (please tell me if that's the case), so we may need a way to configure more precisely the flags/args when running the container, or these arguments should not always be set.

Thanks for your time!

### Expected behavior

The integration test runs successfully (same result as the integration test with a jar / quarkus.container-image.build=false and as the unit test).

### Actual behavior

The integration test fails because the container doesn't start.

### How to Reproduce?

Run the linked project with `mvn clean verify` (or mvnw)
[quarkus-podman-issue.zip](https://github.com/user-attachments/files/21427548/quarkus-podman-issue.zip)

### Output of `uname -a` or `ver`

Linux lxdev1 5.14.0-162.23.1.el9_1.x86_64 #1 SMP PREEMPT_DYNAMIC Thu Mar 23 20:08:28 EDT 2023 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "21.0.1" 2023-10-17 OpenJDK Runtime Environment (build 21.0.1+12-29) OpenJDK 64-Bit Server VM (build 21.0.1+12-29, mixed mode, sharing)

### Quarkus version or git rev

3.25.0

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.9 (8e8579a9e76f7d015ee5ec7bfcdc97d260186937) Maven home: /opt/apps/pkgs/maven Java version: 21.0.1, vendor: Oracle Corporation, runtime: /usr/lib/jvm/jdk-21.0.1 Default locale: fr_CH, platform encoding: UTF-8 OS name: "linux", version: "5.14.0-162.23.1.el9_1.x86_64", arch: "amd64", family: "unix"

### Additional information

_No response_

--- a/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/ContainerUserResolver.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/ContainerUserResolver.java
@@ -1,0 +1,81 @@
+package io.quarkus.deployment.pkg.steps;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.jboss.logging.Logger;
+
+import io.quarkus.deployment.util.ContainerRuntimeUtil.ContainerRuntime;
+import io.smallrye.common.process.ProcessBuilder;
+import io.smallrye.common.process.ProcessExecutionException;
+
+/**
+ * Resolves the UID and GID of the user that runs inside a container image by executing
+ * {@code id} inside an ephemeral container. Results are cached per image for the lifetime of the JVM
+ * to avoid repeated process launches.
+ */
+public final class ContainerUserResolver {
+
+    private static final Logger log = Logger.getLogger(ContainerUserResolver.class);
+
+    private static final Pattern ID_OUTPUT_PATTERN = Pattern.compile("uid=(\\d+).*gid=(\\d+)");
+
+    // Cache keyed by image name. Only successful resolutions are cached.
+    private static final Map<String, ContainerUser> CACHE = new ConcurrentHashMap<>();
+
+    private ContainerUserResolver() {
+    }
+
+    /**
+     * Returns the {@link ContainerUser} (uid, gid) for the default user of the given image,
+     * or {@code null} if detection failed (e.g. the {@code id} command is not available).
+     * Successful resolutions are cached for the lifetime of the JVM; failures are not cached
+     * so callers can retry after transient errors (e.g. network issues while pulling the image).
+     */
+    public static ContainerUser resolve(ContainerRuntime runtime, String image) {
+        ContainerUser cached = CACHE.get(image);
+        if (cached != null) {
+            return cached;
+        }
+        ContainerUser result = runIdInContainer(runtime, image);
+        if (result != null) {
+            CACHE.put(image, result);
+        }
+        return result;
+    }
+
+    private static ContainerUser runIdInContainer(ContainerRuntime runtime, String image) {
+        try {
+            String output = ProcessBuilder.newBuilder(runtime.getExecutableName())
+                    .arguments("run", "--rm", "--entrypoint", "id", image)
+                    .output().gatherOnFail(true).toSingleString(8192)
+                    .error().redirect()
+                    .run()
+                    .trim();
+
+            Matcher matcher = ID_OUTPUT_PATTERN.matcher(output);
+
+            if (matcher.find()) {
+                return new ContainerUser(matcher.group(1), matcher.group(2));
+            }
+
+            log.warnf("Could not parse 'id' output for image '%s': %s", image, output);
+            return null;
+
+        } catch (ProcessExecutionException e) {
+            log.warnf("Could not resolve container UID/GID for image '%s': %s", image, e.getMessage());
+            return null;
+        } catch (Exception e) {
+            if (e instanceof InterruptedException) {
+                Thread.currentThread().interrupt();
+            }
+            log.warnf(e, "Error resolving container UID/GID for image '%s'", image);
+            return null;
+        }
+    }
+
+    public record ContainerUser(String uid, String gid) {
+    }
+}

--- a/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildLocalContainerRunner.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildLocalContainerRunner.java
@@ -9,6 +9,7 @@ import java.util.Collections;
 import java.util.List;
 
 import org.apache.commons.lang3.SystemUtils;
+import org.jboss.logging.Logger;
 
 import io.quarkus.deployment.pkg.NativeConfig;
 import io.quarkus.deployment.util.ContainerRuntimeUtil.ContainerRuntime;
@@ -16,33 +17,78 @@ import io.quarkus.deployment.util.FileUtil;
 
 public class NativeImageBuildLocalContainerRunner extends NativeImageBuildContainerRunner {
 
+    private static final Logger log = Logger.getLogger(NativeImageBuildLocalContainerRunner.class);
+
     public NativeImageBuildLocalContainerRunner(NativeConfig nativeConfig) {
         super(nativeConfig);
         List<String> containerRuntimeArgs = new ArrayList<>(Arrays.asList(baseContainerRuntimeArgs));
         if (SystemUtils.IS_OS_LINUX && containerRuntime.isInWindowsWSL()) {
             containerRuntimeArgs.add("--interactive");
         }
-        containerRuntimeArgs.addAll(getVolumeAccessArguments(containerRuntime));
+        containerRuntimeArgs
+                .addAll(getVolumeAccessArguments(containerRuntime, nativeConfig.builderImage().getEffectiveImage()));
         baseContainerRuntimeArgs = containerRuntimeArgs.toArray(baseContainerRuntimeArgs);
     }
 
-    public static List<String> getVolumeAccessArguments(ContainerRuntime containerRuntime) {
+    /**
+     * Returns the container runtime arguments needed to ensure that files written to a bind-mounted
+     * volume on the host are owned by the real host user.
+     * <p>
+     * The UID/GID of the running user inside {@code image} is detected dynamically via
+     * {@link ContainerUserResolver} so that any container image works correctly, regardless of
+     * which user it declares. When detection fails, Podman falls back to {@code --user} with the
+     * host UID/GID (and {@code --userns=keep-id} for rootless Podman).
+     *
+     * @param containerRuntime the detected container runtime
+     * @param image the container image that will be run (used for UID/GID detection)
+     * @return the list of extra {@code run} arguments, or an empty list when not applicable
+     */
+    public static List<String> getVolumeAccessArguments(ContainerRuntime containerRuntime, String image) {
         if (containerRuntime.isUnavailable()) {
             return List.of();
         }
 
         final List<String> result = new ArrayList<>();
         if (SystemUtils.IS_OS_LINUX || SystemUtils.IS_OS_MAC) {
-            if (containerRuntime.isDocker() && containerRuntime.isRootless()) {
-                Collections.addAll(result, "--user", String.valueOf(0));
+            if (containerRuntime.isRootless()) {
+                if (containerRuntime.isDocker()) {
+                    Collections.addAll(result, "--user", String.valueOf(0));
+                } else if (containerRuntime.isPodman()) {
+                    // Rootless Podman: map the container user to the host user inside the user namespace.
+                    // The UID/GID is resolved dynamically so any builder image works correctly.
+                    ContainerUserResolver.ContainerUser containerUser = ContainerUserResolver.resolve(containerRuntime, image);
+                    if (containerUser == null) {
+                        String uid = getLinuxID("-ur");
+                        String gid = getLinuxID("-gr");
+                        if (uid != null && gid != null && !uid.isEmpty() && !gid.isEmpty()) {
+                            Collections.addAll(result, "--user", uid + ":" + gid);
+                            result.add("--userns=keep-id");
+                        } else {
+                            log.warn(
+                                    "Cannot determine host UID/GID; rootless Podman volume permissions may be incorrect.");
+                        }
+                    } else {
+                        result.add("--userns=keep-id:uid=" + containerUser.uid() + ",gid=" + containerUser.gid());
+                    }
+                }
             } else {
                 String uid = getLinuxID("-ur");
                 String gid = getLinuxID("-gr");
                 if (uid != null && gid != null && !uid.isEmpty() && !gid.isEmpty()) {
-                    Collections.addAll(result, "--user", uid + ":" + gid);
-                    if (containerRuntime.isPodman() && containerRuntime.isRootless()) {
-                        // Needed to avoid AccessDeniedExceptions
-                        result.add("--userns=keep-id");
+                    if (containerRuntime.isPodman()) {
+                        // Rootful Podman: remap the container user's UID/GID to the host user's UID/GID.
+                        ContainerUserResolver.ContainerUser containerUser = ContainerUserResolver.resolve(containerRuntime,
+                                image);
+                        if (containerUser == null) {
+                            Collections.addAll(result, "--user", uid + ":" + gid);
+                        } else {
+                            // Map system IDs (including root) so bind-mounted files remain accessible
+                            Collections.addAll(result, "--uidmap", "0:0:999", "--gidmap", "0:0:999");
+                            Collections.addAll(result, "--uidmap", containerUser.uid() + ":" + uid + ":1");
+                            Collections.addAll(result, "--gidmap", containerUser.gid() + ":" + gid + ":1");
+                        }
+                    } else {
+                        Collections.addAll(result, "--user", uid + ":" + gid);
                     }
                 }
             }

--- a/test-framework/common/src/main/java/io/quarkus/test/common/DefaultDockerContainerLauncher.java
+++ b/test-framework/common/src/main/java/io/quarkus/test/common/DefaultDockerContainerLauncher.java
@@ -123,7 +123,7 @@ public class DefaultDockerContainerLauncher implements DockerContainerArtifactLa
             args.add("--rm");
 
             if (!volumeMounts.isEmpty()) {
-                args.addAll(NativeImageBuildLocalContainerRunner.getVolumeAccessArguments(containerRuntime));
+                args.addAll(NativeImageBuildLocalContainerRunner.getVolumeAccessArguments(containerRuntime, containerImage));
             }
 
             if (httpPort != 0) {
@@ -240,7 +240,7 @@ public class DefaultDockerContainerLauncher implements DockerContainerArtifactLa
         args.add("--rm");
 
         if (!volumeMounts.isEmpty()) {
-            args.addAll(NativeImageBuildLocalContainerRunner.getVolumeAccessArguments(containerRuntime));
+            args.addAll(NativeImageBuildLocalContainerRunner.getVolumeAccessArguments(containerRuntime, containerImage));
         }
 
         args.add("-p");
@@ -431,7 +431,7 @@ public class DefaultDockerContainerLauncher implements DockerContainerArtifactLa
 
         ContainerRuntime containerRuntime = ContainerRuntimeUtil.detectContainerRuntime();
         if (!volumeMounts.isEmpty()) {
-            args.addAll(NativeImageBuildLocalContainerRunner.getVolumeAccessArguments(containerRuntime));
+            args.addAll(NativeImageBuildLocalContainerRunner.getVolumeAccessArguments(containerRuntime, containerImage));
         }
 
         args.addAll(toEnvVar("JAVA_TOOL_OPTIONS",

